docs/patterns.md

```diff
@@ -142,3 +142,131 @@ Example access policy:
 * The front desk cannot cut releases nor deploy services
 
 Complex policies like this are currently best implemented in code directly, though there is [ongoing work](https://github.com/michaelansel/hubot-rbac) to build a generalized framework for access management.
+
+## Middleware Examples
+
+Middleware is often used for authentication and authorization.  There are three kinds of middleware:
+- **Receive**: executed for every message prior to testing for listener matches
+- **Listener**: executed for every listener that matches a message before executing the listener callback
+- **Response**: executed for every message hubot attempts to send, regardless of reason or source
+
+Middleware is asynchronous by default, needing to call either `middleware.next()` to continue processing or `middleware.done()` to stop processing. There are Sync helper functions for each type to simplify creating simple middleware units.
+
+**All of these are just examples. None of them are secure. Don't use them.**
```

Comment:
I'm somewhat dredging this out of the depths of my memory, but I think the big issues here was having a good understanding of your threat model and the list of all listeners. Blacklisting is particularly dangerous if you are auto-updating scripts and the upstream changes the listener ID (it is suddenly allowed with the same regex).

I ended up rewriting this section when we actually merged ("Restricting access to commands") -- https://hubot.github.com/docs/patterns/